cerebro: init at 0.11.0

kylesferrazza opened the pull request:
###### Description of changes

Created the cerebro derivation.
Based on [`todoist-electron`](https://github.com/NixOS/nixpkgs/blob/5c90ea3fa6841484fffde6c04d04461a69ae7582/pkgs/applications/misc/todoist-electron/default.nix).

Closes #214977 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->
Made sure it runs and is able to locate and run installed apps, and search the internet.

dotlambda reviewed (dismissed):
Specify `meta.sourceProvenance`.

kylesferrazza commented:
> Specify `meta.sourceProvenance`.

Done

erkkimon commented:
First of all, thanks @kylesferrazza for your contribution! Is there any blocker here? All checks have passed. We need Cerebro. :) ping @dotlambda

eclairevoyant commented on pkgs/applications/misc/cerebro/default.nix:
```diff
@@ -0,0 +1,53 @@
+
+  meta = with lib; {
+    homepage = "https://www.cerebroapp.com/";
+    description = "An open-source launcher to improve your productivity and efficiency";
```
```suggestion
    description = "Open-source launcher to improve your productivity and efficiency";
```

https://github.com/NixOS/nixpkgs/blob/b59dfac3c65d7ccc12426527553c9250cb71e90a/pkgs/README.md?plain=1#L375-L378

eclairevoyant commented on pkgs/applications/misc/cerebro/default.nix:
```diff
@@ -0,0 +1,53 @@
+
+  src = fetchurl {
+    url = "https://github.com/cerebroapp/cerebro/releases/download/v${version}/Cerebro-${version}.AppImage";
+    sha256 = "sha256-+rjAMoQI3KTmHGFnyFoe20qIrAEi0DL3ksInFy677P8=";
```
```suggestion
    hash = "sha256-+rjAMoQI3KTmHGFnyFoe20qIrAEi0DL3ksInFy677P8=";
```

eclairevoyant commented on pkgs/applications/misc/cerebro/default.nix:
```diff
@@ -0,0 +1,53 @@
+  };
+
+  appimageContents = appimageTools.extractType2 {
+    name = "${pname}-${version}";
+    inherit src;
```
```suggestion
    inherit pname version src;
```

eclairevoyant commented on pkgs/applications/misc/cerebro/default.nix:
```diff
@@ -0,0 +1,53 @@
+      --add-flags "\''${NIXOS_OZONE_WL:+\''${WAYLAND_DISPLAY:+--ozone-platform-hint=auto --enable-features=WaylandWindowDecorations}}"
+  '';
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

tracking issue: #292468

eclairevoyant commented on pkgs/applications/misc/cerebro/default.nix:
```diff
@@ -0,0 +1,53 @@
+    homepage = "https://www.cerebroapp.com/";
+    description = "An open-source launcher to improve your productivity and efficiency";
+    platforms = [ "x86_64-linux" ];
+    license = licenses.mit;
+    maintainers = with maintainers; [ kylesferrazza ];
+    sourceProvenance = with sourceTypes; [ binaryNativeCode ];
```
```suggestion
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ kylesferrazza ];
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
```

eclairevoyant replied:
also, please set `meta.mainProgram`

eclairevoyant commented on pkgs/top-level/all-packages.nix:
```diff
@@ -398,6 +398,8 @@ with pkgs;
 
   cereal = callPackage ../development/libraries/cereal { };
 
+  cerebro = callPackage ../applications/misc/cerebro { };
```
please move to `pkgs/by-name`, and remove this line

eclairevoyant commented on pkgs/applications/misc/cerebro/default.nix:
```diff
@@ -0,0 +1,53 @@
+{ lib, stdenv, fetchurl, appimageTools, makeWrapper, electron_21, libsecret }:
```
electron 21 has been removed, any chance we can use a current electron?

eclairevoyant reviewed (changes requested):
Thanks for the contribution. There's some changes to be made below, as the packaging guidelines have changed since the original submission.

SuperSandro2000 commented on pkgs/applications/misc/cerebro/default.nix:
```diff
@@ -0,0 +1,53 @@
+
+    mkdir -p $out/bin $out/share/${pname} $out/share/applications $out/share/icons/hicolor/512x512
+
+    cp -a ${appimageContents}/{locales,resources} $out/share/${pname}
+    cp -a ${appimageContents}/cerebro.desktop $out/share/applications/${pname}.desktop
+    cp -a ${appimageContents}/usr/share/icons/hicolor/512x512/apps $out/share/icons/hicolor/512x512
+
+    substituteInPlace $out/share/applications/${pname}.desktop \
+      --replace 'Exec=AppRun' 'Exec=${pname}'
```
Please don't use pname wherever possible

kylesferrazza commented:
Addressed all the feedback, and all appears to go well, but unfortunately seeing a blank window when I launch the application. I'll need to keep investigating and push a fix. Converted to draft for now.

kylesferrazza commented:
Could be `nvidia` related... 
```bash
$ ./result/bin/cerebro 

(electron:22952): Gtk-WARNING **: 13:22:27.639: Theme parsing error: gtk.css:9672:21: Not using units is deprecated. Assuming 'px'.
[22952:0825/132227.696167:ERROR:atom_cache.cc(229)] Add _NET_WM_WINDOW_TYPE_SPLASH to kAtomsToCache
DRM kernel driver 'nvidia-drm' in use. NVK requires nouveau.
```

eclairevoyant commented:
Also as per CI:

```
Please go to the Nixpkgs root directory, run `nix-shell`, then:
nixfmt 'pkgs/by-name/ce/cerebro/package.nix'
```

kylesferrazza committed:
cerebro: init at 0.11.0
